### PR TITLE
Pass durations to backend

### PR DESCRIPTION
**Pull request by vegaro**
Adds subscripion period to the POST receipt request. 

Changes summary: 

- added the fields from https://github.com/RevenueCat/khepri/pull/1539 to the post receipt request:
    - `normal_duration`
    - `intro_duration`
    - `trial_duration`
- Reduced amount of arguments in the postReceipt call by making a `ProductInfo` class.
- Some small refactor on tests in PurchasesTests.
- I created a new `PostingTransactionsTests.kt`. I will migrate some post receipt tests over there in the near future.

### CODE REVIEW

File: purchases/src/test/java/com/revenuecat/purchases/BackendTest.kt [hunk 1]
```diff
@@ -570,57 +574,59 @@ class BackendTest {
...
-            null
+            isRestore = false,
+            responseCode = 200,
+            clientException = null,
+            resultBody = null,
+            delayed = true,
+            offeringIdentifier = "offering_b",
+            observerMode = false,
+            price = null,
+            currency = null
```
Comment (aboedo) [inline purchases/src/test/java/com/revenuecat/purchases/BackendTest.kt:613]: love all the additions of named arguments

File: purchases/src/test/java/com/revenuecat/purchases/PostingTransactionsTests.kt [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+import io.mockk.verify
+import org.assertj.core.api.Assertions.assertThat
+import org.json.JSONObject
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import java.util.concurrent.ExecutorService
+
+@RunWith(AndroidJUnit4::class)
+class PostingTransactionsTests {
```
Comment (aboedo) [inline purchases/src/test/java/com/revenuecat/purchases/PostingTransactionsTests.kt:21]: awesome

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 3]
```diff
@@ -907,17 +891,23 @@ class Purchases @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE) intern
...
         val unsyncedSubscriberAttributesByKey =
             subscriberAttributesManager.getUnsyncedSubscriberAttributes(appUserID)
+        val productInfo = ProductInfo(
+            productID = purchase.sku,
+            offeringIdentifier = purchase.presentedOfferingIdentifier,
+            price = skuDetails?.priceAmount,
+            currency = skuDetails?.priceCurrencyCode,
+            duration = skuDetails?.subscriptionPeriod?.takeUnless { it.isEmpty() },
+            introDuration = skuDetails?.introductoryPricePeriod?.takeUnless { it.isEmpty() },
+            trialDuration = skuDetails?.freeTrialPeriod?.takeUnless { it.isEmpty() }
```
Comment (aboedo) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:901]: it looks like we're getting a lot of data from skuDetails, maybe ProductInfo could have a constructor that takes skuDetails and read from there?

File: purchases/src/test/java/com/revenuecat/purchases/PostingTransactionsTests.kt [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            allowSharingPlayStoreAccount = true,
+            consumeAllTransactions = true,
+            appUserID = appUserId,
+            onSuccess = {_,_ -> },
+            onError = {_,_ -> }
+        )
+        assertThat(postedProductInfoSlot.isCaptured).isTrue()
+        assertThat(postedProductInfoSlot.captured.duration).isEqualTo("P1M")
+        assertThat(postedProductInfoSlot.captured.introDuration).isEqualTo("P2M")
+        assertThat(postedProductInfoSlot.captured.trialDuration).isEqualTo("P3M")
```
Comment (aboedo) [inline purchases/src/test/java/com/revenuecat/purchases/PostingTransactionsTests.kt:134]: FFTI these strings could be extracted

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 3 above)
Comment (vegaro) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:901]: I love the idea. Easier to test stuff too

**Comment by vegaro** [comment]
I am going to hold the changes on adding the `SKUDetails` to the `ProductInfo` because it's breaking too many tests and will slow the release. But I promise it will get done.

**Comment by aboedo** [comment]
No worries, I figured that it could happen. We can take care of it next week.

**Pull request merged**